contracts/staking/DelegationManager.sol

```diff
@@ -1,230 +1,298 @@
+    uint256 currentEpoch = stakeManager.currentEpoch();
+
+    if (delegators[delegatorId].bondedTo != 0) {
+      _unbond(delegatorId, currentEpoch, stakeManager);
+    }
+
+    require(delegators[delegatorId].deactivationEpoch == 0);
+    delegators[delegatorId].deactivationEpoch = currentEpoch.add(WITHDRAWAL_DELAY);
+    emit UnstakeInit(msg.sender, delegatorId, delegators[delegatorId].deactivationEpoch);
+  }
+
+  // after unstaking wait for WITHDRAWAL_DELAY, in order to claim stake back
+  function unstakeClaim(
+    uint256 delegatorId,
+    uint256 accumBalance,
+    uint256 accumSlashedAmount,
+    uint256 accIndex,
+    bytes memory accProof
+    ) public onlyDelegator(delegatorId) {
+    IStakeManager stakeManager = IStakeManager(registry.getStakeManagerAddress());
+    require(
+      keccak256(
+        abi.encodePacked(
+          delegatorId,
+          accumBalance,
+          accumSlashedAmount)
+          ).checkMembership(
+            accIndex,
+            stakeManager.accountStateRoot(),
+            accProof),
+      "Wrong account proof"
+      );
     Delegator storage delegator = delegators[delegatorId];
+    require(
+      delegators[delegatorId].deactivationEpoch > 0 &&
+      delegators[delegatorId].deactivationEpoch.add(
+      stakeManager.WITHDRAWAL_DELAY()) <= stakeManager.currentEpoch(),
+      "Incomplete withdraw Period"
+      );
+
+    uint256 _reward = accumBalance.sub(delegator.claimedRewards);
+    uint256 _slashedAmount = accumSlashedAmount.sub(delegator.slashedAmount);
+
+    require(stakeManager.updateTotalRewardsLiquidated(_reward), "Liquidating more rewards then checkpoints submitted");
+
+    uint256 amount = delegator.amount.sub(_slashedAmount);
+    totalStaked = totalStaked.sub(delegator.amount);
+
+    //@todo :add slashing, take slashedAmount into account for totalStaked
+    stakerNFT.burn(delegatorId);
+    // @todo merge delegationManager/stakeManager capital and rewards
+    require(stakeManager.delegatorWithdrawal(delegator.reward.add(_reward), stakerNFT.ownerOf(delegatorId)),"Amount transfer failed");
+    require(token.transfer(msg.sender, amount));
+    delete delegators[delegatorId];
+    emit Unstaked(msg.sender, delegatorId, amount, totalStaked);
+  }
 
-    require(delegator.lastValidatorEpoch == 0 || delegator.lastValidatorEpoch.add(validatorHopLimit) <= currentEpoch, "Delegation_Limit_Reached");
+  function bond(uint256 delegatorId, uint256 validatorId) public onlyDelegator(delegatorId) {
+    IStakeManager stakeManager = IStakeManager(registry.getStakeManagerAddress());
+    uint256 currentEpoch = stakeManager.currentEpoch(); //TODO add 1
 
-    address validator;
-    (,,,,,,,validator,) = stakeManager.validators(validatorId);
-    require(validator != address(0x0), "Unknown validatorId or validator doesn't expect delegations");
+    if (delegators[delegatorId].bondedTo != 0) {
+      emit ReBonding(delegatorId, delegators[delegatorId].bondedTo, validatorId);
```

Comment:
After ReBonding event, we are also emitting Bonding Event. 
pls remove Bonding event after rebonding event.

Reply:
ignore the Rebonding event, I've kept it for the outside world. on heimdall just use bond/unbond

Reply:
That would be 2 transactions on heimdall which are unnecessary. 
Let's keep Rebond event

Reply:
okey makes sense made changes